Add blog post around Hacktoberfest prep with dependency analysis (#263)
deliveroo/deliveroo.engineering: _posts/2022-09-28-hacktoberfest-dependency-analysis.md

---
layout: post
title: Analysing our dependency trees to determine where we should send Open Source contributions for Hacktoberfest
authors:
  - "Jamie Tanna"
excerpt: >
  How we're using GitHub Advanced Security's dependency scanning
  functionality to determine what our most popular dependencies are, and whether
  we can find any Open Source contributions for the month of Hacktoberfest.

> date: 2022-09-28 # TODO

Author: I'll update this and the filename before we ship it 🚀

---
## Table of Contents
{:.no_toc}

1. Automatic Table of Contents Here
{:toc}

## Why contributing to Open Source is so important

> In recent years, it's become unavoidable to build software on top of Open Source. This is _absolutely_ a great thing, and allows developers to focus on fewer areas of domain specialisation as possible, as well as allowing a much wider range of users to pick up on defects and bring new features to our tools.

Reviewer: ```suggestion
In recent years, it has become unavoidable to build software on top of Open Source. This is _absolutely_ a great thing, and allows developers to focus on fewer areas of domain specialisation as possible, as well as allowing a much wider range of users to pick up on defects and bring new features to our tools.
```

However, with events such as the [Log4Shell](https://en.wikipedia.org/wiki/Log4Shell) security vulnerability, times where maintainers have removed their libraries [from package and source repositories](https://www.bleepingcomputer.com/news/security/dev-corrupts-npm-libs-colors-and-faker-breaking-thousands-of-apps/), sometimes in [political protest](https://www.wired.com/story/developer-deletes-code-protest-ice/), it's understandable that businesses are somewhat hesitant about the sustainability of projects.

One of the key things that all Open Source licences highlight is that they're provided "as is", with “no warranty”. Although maintainers may get to that bug fix, or incredibly high severity security fix, the whole point is that the source is available for you, dear reader, to contribute that fix without needing to wait for a maintainer to assist you.

> Sometimes it's straightforward to spot an issue and try to fix it yourself, but if you've never contributed to an Open Source project before, it can be particularly daunting, especially if it's a considerably large codebase, or if it's critical infrastructure for several companies and projects.

Reviewer: ```suggestion
Sometimes it's straightforward to spot an issue and try to fix it yourself, but if you've never contributed to an Open Source project before it can be particularly daunting, especially if it's a considerably large codebase or if it's critical infrastructure for several companies and projects.
```

> Contributing to Open Source is one of those things that's best learned by doing, which is fortunate because in the month of October, DigitalOcean are running an event across the world run called [Hacktoberfest](https://hacktoberfest.com), which aims to be a celebration of Free and Open Source and to give more folks exposure to contributing to projects. I've been participating in Hacktoberfest since 2014, and it has truly changed the way that I engage with Open Source. It's given me the opportunity to be more comfortable engaging in Open Source, and as a maintainer on several large projects myself, it's given me further empathy on both sides of the experience.

Reviewer: ```suggestion
Contributing to Open Source is one of those things that's best learned by doing, which is fortunate because in the month of October, DigitalOcean are running an event across the world run called [Hacktoberfest](https://hacktoberfest.com), which aims to be a celebration of Free and Open Source and to give more folks exposure to contributing to projects. I've been participating in Hacktoberfest since 2014, and it has truly changed the way that I engage with Open Source. It has given me the opportunity to be more comfortable engaging in Open Source and, as a maintainer on several large projects myself, it has given me further empathy on both sides of the experience.
```

A few months ago, I [wrote about an idea on my personal blog](https://www.jvt.me/posts/2022/06/01/idea-supply-chain-monetisation/) about programmatically determining how (Open Source) libraries are used and, in that case, contributing financially, but the concept still works for contributing in other ways. I decided that I wanted to use the same dependency analysis approach, using the dependency tracking functionality we have available through GitHub Advanced security. Deliveroo is a data-driven company, so being able to bring some data to teams, to highlight commonly used libraries that may be good candidates for contributions, was really important.

To do this, the overarching steps were:

- get a list of all repositories to scan
- get dependency scanning enabled across all of our repositories, which was scriptable
- grab the list of dependencies for each repository
- extract the dependencies for each repo and insert them into an SQLite database
- run some queries to pull out the most popular direct and transitive dependencies

> Across our repositories, it's been especially interesting to see that our most used (direct) dependencies seem to be testing, data serialisation/deserialisation, and frameworks like React and Rails.

Reviewer: ```suggestion
Across our repositories, its been especially interesting to see that our most used (direct) dependencies seem to be testing, data serialisation/deserialisation, and frameworks like React and Rails.
```
